### DISCUSSION TITLE
Add item to tracestate

### DISCUSSION THREAD

**Discussion by knutwannheden** [open-telemetry/opentelemetry-java#4008] [Q&A]
I have a use case where I want a service to add or modify an item in the trace state (deserialized from an incoming W3C `tracestate` header) so that all future child spans which get created are associated with a context with this modified trace state.

I could supply more details regarding the use case if you think this sounds like a bad idea. But I would have expected this use case to be supported. I could however not find any simple solution. The only solution I found so far was to replace the textmap propagator for the `tracestate` header.

**Comment by jkwatson** [comment]
Hi @knutwannheden . This should be relatively easy to do with the APIs as-is. The traceState is in the propagated span, which will be put into the Context returned by the propagator. If you get that span out of the context, you can get the TraceState from the SpanContext inside that Span, and then call `toBuilder()` on that TraceState instance, do what you need to do, then construct a new propagated span with that modified trace state and put it into a new Context.

If the APIs aren't obvious (there are a bunch of moving parts), I can probably put together a gist that would show how to do it.

**Reply by nschile** [reply]
I'm also wanting to wanting to add an items to the `tracestate`. ~~I tried the below and it seems to work, though the resulting span isn't exported because `Span.wrap` states the `returned Span will not be exported and all tracing operations are no-op`. Is there a way to create a Span with the addition of a new item in tracestate that is exportable?~~ I was mistaken, The below code exports the span as expected.

```
SpanContext parentSpanContext = Span.current().getSpanContext();
TraceState traceState = parentSpanContext.getTraceState().toBuilder().put("my-span-id", "123").build();
SpanContext context = SpanContext.create(
        parentSpanContext.getTraceId(), 
        parentSpanContext.getSpanId(), 
        parentSpanContext.getTraceFlags(), 
        traceState);

tracer.spanBuilder("my span name")
        .setParent(Context.current().with(Span.wrap(context)))
        .startSpan();
```

**Reply by jack-berg** [reply]
In your example the span from `Span.current().getSpanContext()` will still be exported - it just won't include the updated `TraceState`. Are you saying that it's important for the exported span's `TraceState` to reflect the updated value?

**Reply by nschile** [reply]
I've updated the comment above. The above code will export the span as expected.

**Reply by przemos** [reply]
hi, how to get hold of the `tracer` ?

**Reply by kpatrick** [reply]
nschile, could you please explain why it is exported?  The docs seem to indicate otherwise.  Are you confusing export with propagate?
